core/domain/fs_domain.py

```diff
@@ -521,43 +531,68 @@ def commit(self, unused_user_id, filepath, raw_bytes, mimetype):
         gcs_file.write(raw_bytes)
         gcs_file.close()
 
-    def delete(self, user_id, filepath):
-        """Raises NotImplementedError if the method is not implemented in the
-        derived classes.
+    def delete(self, user_id, filepath):  # pylint: disable=unused-argument
+        """Deletes a file and the metadata associated with it.
+
+        If `user_id` argument is unused. It is included so that this method
+        signature matches that of other file systems.
 
         Args:
-            user_id: str. The id of the user.
+            user_id: str. The user_id of the user who wants to create or update
+                a file.
             filepath: str. The path to the relevant file within the exploration.
-
-        Raises:
-            NotImplementedError. The method is not implemented in the derived
-                classes.
         """
-        raise NotImplementedError
+        bucket_name = app_identity_services.get_gcs_resource_bucket_name()
+        gcs_file_url = (
+            '/%s/%s/assets/%s' % (
+                bucket_name, self._exploration_id, filepath))
+        try:
+            cloudstorage.delete(gcs_file_url)
+        except cloudstorage.NotFoundError:
+            raise IOError('Image does not exist: %s' % filepath)
+
 
-    def listdir(self, dir_name):
+    def listdir(self, dir_name, delimiter=None):  # pylint: disable=unused-argument
         """Lists all files in a directory.
 
+        The delimiter argument can be used to restrict the results to only the
+        files in the given folder. Without the delimiter, the entire tree
+        under the prefix is returned. For example, given these blobs:
+        /a/1.txt
+        /a/b/2.txt
+        If you just specify prefix = '/a', you'll get back:
```

Comment:
Oh, I see. I think this is OK, you know. That's what listdir() does: it would list all files and folders. What you should do is to just run over the results in your job and make sure none of them contain a slash.

Keeping this example here is a good idea. But I would get rid of delimiters entirely and avoid changing the behaviour. Modify your job instead to verify that there are no subdirs.

Reply:
Okay I will remove the delimiter part from the fs_domain.
We have a validation job right?. That already validates the url, and the REGEX is such that selects the url such that they follow this
/image/{{name}}.{{extension}} 
Regex
`'^/([^/]+)/([^/]+)/assets/image/(([^/]+)\.(' + '|'.join(
        ALLOWED_IMAGE_EXTENSIONS) + '))$'`
Hence it already makes sure that after `image/` there are no subdirectories. And we list the directories with the prefix till `image/` So the case with subdirectories is not possible.

Reply:
Yep, sounds good then. Thanks!

Reply:
Done